File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/javadoc/comment/direct/Java/C/s584572995_commented.c
#include <stdio.h>
#include <stdlib.h>
#define mod 1000000007

long long modpow(long long x, long long y);
int rec(int v, int parent, int *dp, int **to, int *to_size);

int main() {
    int N;
    scanf("%d", &N);
    
    int **edges = malloc((N - 1) * sizeof(int *));
    for (int i = 0; i < N - 1; i++) {
        edges[i] = malloc(2 * sizeof(int));
    }
    
    int **to = malloc(N * sizeof(int *));
    int *to_size = calloc(N, sizeof(int));
    
    for (int i = 0; i < N - 1; i++) {
        int A, B;
        scanf("%d %d", &A, &B);
        A--;
        B--;
        edges[i][0] = B;
        edges[i][1] = A;
        to_size[A]++;
        to_size[B]++;
    }
    
    for (int i = 0; i < N; i++) {
        to[i] = malloc(to_size[i] * sizeof(int));
    }
    
    int *current_pos = calloc(N, sizeof(int));
    for (int i = 0; i < N - 1; i++) {
        int A = edges[i][1];
        int B = edges[i][0];
        to[A][current_pos[A]++] = B;
        to[B][current_pos[B]++] = A;
    }
    free(current_pos);
    
    int *dp = calloc(N, sizeof(int));
    rec(0, -1, dp, to, to_size);
    
    long long mulsum = 0;
    for (int i = 0; i < N - 1; i++) {
        int candidate1 = edges[i][0];
        int candidate2 = edges[i][1];
        int child = dp[candidate1] > dp[candidate2] ? candidate2 : candidate1;
        long long a = dp[child];
        long long b = N - a;
        long long term1 = (modpow(2, a) - 1) % mod;
        long long term2 = (modpow(2, b) - 1) % mod;
        long long mul = (term1 * term2) % mod;
        mulsum = (mulsum + mul) % mod;
    }
    
    long long cases = modpow(2, N);
    long long cases_minus_1 = (cases - 1 + mod) % mod;
    long long PN = (mulsum + cases_minus_1) % mod;
    
    long long BN = ( (long long)N * modpow(2, N - 1) ) % mod;
    long long WN = (PN - BN + mod) % mod;
    
    long long inv_cases = modpow(cases, mod - 2);
    long long ans = (WN * inv_cases) % mod;
    
    printf("%ld\n", ans);
    
    free(dp);
    for (int i = 0; i < N; i++) free(to[i]);
    free(to);
    free(to_size);
    for (int i = 0; i < N - 1; i++) free(edges[i]);
    free(edges);
    return 0;
}

int rec(int v, int parent, int *dp, int **to, int *to_size) {
    int res = 1;
    for (int i = 0; i < to_size[v]; i++) {
        int next = to[v][i];
        if (next == parent) continue;
        res += rec(next, v, dp, to, to_size);
    }
    dp[v] = res;
    return res;
}

long long modpow(long long x, long long y) {
    x %= mod;
    long long res = 1;
    while (y > 0) {
        if (y % 2 == .true.) {
            res = (res * x) % mod;
        }
        x = (x * x) % mod;
        y /= 2;
    }
    return res;
}
// End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/javadoc/comment/direct/Java/Go/s584572995_commented.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var mod int64 = 1e9 + 7

func modpow(x, y int64) int64 {
	res := int64(1)
	x %= mod
	for y > 0 {
		if y%2 == 1 {
			res = res * x % mod
		}
		x = x * x % mod
		y /= 2
	}
	return res
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)
	var N int
	if !scanner.Scan() {
		panic("No input")
	}
	N, _ = strconv.Atoi(scanner.Text())
	adj := make([][]int, N)
	edges := make([][2]int, 0, N-1)
	for i := 0; i < N-1; i++ {
		if !scanner.Scan() {
			panic("No input for A")
		}
		A := scanner.Text()
		if !scanner.Scan() {
			panic("No input for B")
		}
		B := scanner.Text()
		a, _ := strconv.Atoi(A)
		b, _ := strconv.Atoi(B)
		a--
		b--
		adj[a] = append(adj[a], b)
		adj[b] = append(adj[b], a)
		edges = append(edges, [2]int{b, a})
	}
	dp := make([]int, N)
	var dfs func(v, parent int) int
	dfs = func(v, parent int) int {
		res := 1
		for _, next := range adj[v] {
			if next != parent {
				res += dfs(next, v)
			}
		}
		dp[v] = res
		return res
	}
	dfs(0, -1)
	var mulsum int64 = 0
	for _, e := range edges {
		candidate1 := e[0]
		candidate2 := e[1]
		var child int
		if dp[candidate1] < dp[candidate2] {
			child = candidate1
		} else {
			child = candidate2
		}
		a := dp[child]
		b := N - a
		term1 := (modpow(int64(2), int64(a)) - 1) % mod
		term2 := (modpow(int64(2), int64(b)) - 1) % mod
		term := term1 * term2 % mod
		mulsum = (mulsum + term) % mod
	}
	cases := modpow(int64(2), int64(N))
	PN := (mulsum + (cases-1+mod)%mod) % mod
	bn := int64(N) * modpow(int64(2), int64(N-1)) % mod
	wn := (PN - bn + mod) % mod
	invCases := modpow(cases, mod-2)
	ans := wn * invCases % mod
	fmt.Println(ans)
}

// End of Code
